### PR TITLE
Log metadata server access for master pods by UID

### PR DESCRIPTION
**Pull request by JonathanSun**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Adds iptables firewall rules which log access to the GCE Metadata Server from master pods based on UID.  Exposes visibility into which components are expected or not expected to require metadata access.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by JonathanSun** [comment]
I signed it

File: cluster/gce/gci/configure.sh [hunk 1]
```diff
@@ -656,4 +688,7 @@ ensure-container-runtime
 # binaries and kube-system manifests
 install-kube-binary-config
 
+# Log all metadata access not from approved processes.
+configure-gce-metadata-fw
```
Comment (immutableT) [inline cluster/gce/gci/configure.sh:692]: What do you think about moving this into configure-helper.sh?
Specifically, integrating this into the [configure-ip-firewall ](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#L99)function.

File: cluster/gce/gci/configure.sh [hunk 2]
```diff
@@ -39,6 +39,38 @@ if curl --help | grep -q -- '--retry-connrefused'; then
...
+  local -r invert="${4:-}"
+
+  # Expand rule action to include relevant option flags.
+  case "${action}" in
+    LOG)
+      action="LOG --log-prefix "${prefix}:" --log-uid --log-tcp-options --log-ip-option"
+      ;;
+  esac
+
+  iptables ${command} OUTPUT -p tcp --dport 80 -d 169.254.169.254 -m owner ${invert} --uid-owner=${METADATA_SERVER_ALLOWED_UID_RANGE:-0-2999} -j ${action}
```
Comment (immutableT) [inline cluster/gce/gci/configure.sh:61]: This may raise an error if `invert` is not set; use `${invert:-}`

File: cluster/gce/gci/configure.sh [hunk 3]
```diff
@@ -39,6 +39,38 @@ if curl --help | grep -q -- '--retry-connrefused'; then
...
+}
+
+function configure-gce-metadata-fw {
```
Comment (immutableT) [inline cluster/gce/gci/configure.sh:64]: Optional: consider adding unit tests, see an [example](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/apiserver_etcd_test.go) here.

**Comment by immutableT** [comment]
Please, squash your commits.

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -148,6 +170,15 @@ function config-ip-firewall {
     echo "Add rule for metadata concealment"
     iptables -w -t nat -I PREROUTING -p tcp -d 169.254.169.254 --dport 80 -m comment --comment "metadata-concealment: bridge traffic to metadata server goes to metadata proxy" -j DNAT --to-destination 127.0.0.1:988
```
Comment (mml) [inline cluster/gce/gci/configure-helper.sh:171]: Can we change this line to also use the function above?  Maybe there's a more general function for this?

Or maybe we could at least move 169.254.169.254 to a variable?  The function added is much longer than just copy/pasting the two iptables invocations, and with no other callers it doesn't seem like it's worth the extra complexity.

**Review by mml — CHANGES_REQUESTED** [review]

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (JonathanSun) [reply cluster/gce/gci/configure-helper.sh]: I went the variable route.  There's enough differences between the commands that I don't think generalizing a function is much better than just calling iptables directly.  But the idea of the separate new function was that more similar invocations would be added in the future.

File: cluster/gce/gci/configure-helper.sh [hunk 5]
```diff
@@ -25,6 +25,9 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+### Hardcoded constants
+METADATA_SERVER_IP="169.254.169.254"
```
Comment (karan) [inline cluster/gce/gci/configure-helper.sh:29]: It's flexible to allow overriding but default to the hardcoded IP:

```
METADATA_SERVER_IP="${METADATA_SERVER_IP:-169.254.169.254}"
```

**Pull request merged**